lib: Add date_time library (#1861)
NordicPlayground/fw-nrfconnect-nrf: include/date_time.rst

.. _lib_date_time:

> Date Time Library

Reviewer: ```suggestion
Date-Time Library
```

####################

> The date_time library maintains the current date time UTC from multiple time sources. Every :option:`CONFIG_DATE_TIME_UPDATE_INTERVAL` the library fetches time in a prioritized order as follows:

Reviewer: ```suggestion
The date-time library maintains the current date time information in UTC format after fetching the information from multiple sources.
The option :option:`CONFIG_NRF9160_TIME_UPDATE_INTERVAL` determines the frequency with which the library fetches the time information.
The information is fetched in the following prioritized order:
```

> 1. Check if the age of the current date time is newer than the :option:`CONFIG_DATE_TIME_UPDATE_INTERVAL`, this avoids unnecessary update cycles if the date time already obtained is relatively new. If this check passes a fetch of a new date time is not carried out.

Reviewer: Made some changes here (ref:offline discussion)...
```suggestion
1. The library checks if the time elapsed from the last update to the current date time in UTC format is within the update interval set with the :option:`CONFIG_NRF9160_TIME_UPDATE_INTERVAL`.
    The library does not fetch new date time information if the above check passes.
    In this way, unnecessary update cycles are avoided.

```

> #. If the aforementioned checks fails the library will request time from from the nRF9160s onboard modem.

Reviewer: ````suggestion
#. If the aforementioned check fails, i.e. the time elapsed from the last update to the current date time exceeds the update interval set by :option:`CONFIG_NRF9160_TIME_UPDATE_INTERVAL`, the library requests time from the onboard modem of nRF9160.
````

> #. If the time from the onboard modem is not valid the library will request time from NTP servers. The library will fallback on several NTP servers if a NTP time request does not succeed.

Reviewer: ```suggestion
#. If the time information obtained from the onboard modem of nRF9160 is not valid, the library requests time from NTP servers. 
#. If the NTP time request does not succeed, the library requests time information from several other NTP servers.
```

> The :cpp:func:`date_time_set` function can be used to pass in date time from external sources independent of the internal date time update routine.

Reviewer: ```suggestion
The :cpp:func:`date_time_set` function can be used to obtain the current date time information from external sources independent of the internal date time update routine.
```

> To get date time from the library either call the :cpp:func:`date_time_uptime_to_unix_time_ms` or :cpp:func:`date_time_now` functions. For more information regarding these function refer to the API documentation.

Reviewer: ```suggestion
To get date time information from the library, either call the :cpp:func:`date_time_uptime_to_unix_time_ms` function or the :cpp:func:`date_time_now` function.
See the API documentation for more information on these functions.
````

Configuration
*************

:option:`CONFIG_DATE_TIME_UPDATE_INTERVAL`

>   Configure this option to control how often the library will fetch time.

Reviewer: ```suggestion
  Configure this option to control the frequency with which the library fetches the time information.
```
